[WIP] Setting client side wait_timeout

yashwantsahu20 opened the pull request:
Setting client side wait_timeout – current ProxySQL ignores wait_timeout set by the client. If a client specifies a value for wait_timeout , ProxySQL should terminate the client connection (but not the backend connection) when said timeout is reached instead of ProxySQL’s global mysql-wait_timeout . As a further enhancement, the wait timeout specified by the client shouldn’t exceed ProxySQL’s wait_timeout . The implementation also requires adequate TAP test to verify the functionality.

yashwantsahu20 committed:
Parsing and setting wait timeout for session

renecannao commented:
Brainstorming on implementation details.
Should we perform range validation of wait_timeout specified by the client and return an error if out of range? Or maybe it is easier to silently ignore out of range values? (or maybe write an error in error log without notifying the client).

The previous point is also related to: what happens when we change `mysql-wait_timeout` at runtime?
Currently, the current `mysql-wait_timeout` is what is used to determine if the client needs to be disconnected, not the value of `mysql-wait_timeout` when the client connected. This is used during failover for example, to terminate all clients quickly.
So perhaps we should use the smaller of the two , avoiding out of range validation.

Thoughts?

yashwantsahu20 commented:
Thanks René for very nice suggestions.

Please let me know your thought about this approach.

We can do as following

1. Let's do a basic range validation as done by MySQL to satisfy third-party tools, tests, etc., and stay aligned with MySQL limits.
Only minimum and maximum values will be validated. This will return an error to the client.
Example:
  min - 1
  max - one year

https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_wait_timeout

2. Add an error message to the error log if the session's wait_timeout value is greater than the global one, but still accept and store the value, as the global timeout may be increased in the future.   We need to log a message as client may expect longer session timeout while terminating it.

No need to return an error to the client, since from the client’s perspective, there is no visible failure.

3. While terminating the connection, we will always pick the smaller of the two values (session wait_timeout and global mysql-wait_timeout).

renecannao commented:
SGTM

yashwantsahu20 committed:
Now taking effective timeout taking as minimum of global timeout and session timeout

renecannao commented on lib/MySQL_Session.cpp:
```diff
@@ -6414,6 +6415,54 @@ bool MySQL_Session::handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_C
+									 "wait_timeout must be between %llu ms (1s) and %llu ms (365d), received: %llu ms",
+									 MYSQL_WAIT_TIMEOUT_MIN_MS, MYSQL_WAIT_TIMEOUT_MAX_MS, client_timeout);
+							client_myds->DSS = STATE_QUERY_SENT_NET;
+							client_myds->myprot.generate_pkt_ERR(true, nullptr, nullptr, 1, 1231, (char *)"42000", errmsg, true);
```
This is different from what you suggested yourself:

> No need to return an error to the client, since from the client’s perspective, there is no visible failure.

yashwantsahu20 replied:
This was point 1. But I have checked with MySQL directly , they also dont do any range check. SO I have removed this.

renecannao commented on lib/MySQL_Session.cpp:
```diff
@@ -6414,6 +6415,54 @@ bool MySQL_Session::handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_C
+						if (!is_within_range(client_timeout, MYSQL_WAIT_TIMEOUT_MIN_MS, MYSQL_WAIT_TIMEOUT_MAX_MS)) {
+							char errmsg[256];
+							snprintf(errmsg, sizeof(errmsg),
+									 "wait_timeout must be between %llu ms (1s) and %llu ms (365d), received: %llu ms",
```
No need to hardcode 1s and 365d , as you are defining the values in MYSQL_WAIT_TIMEOUT_MIN_MS and MYSQL_WAIT_TIMEOUT_MAX_MS .
But this is not relevant if we don't return any error to the client.

yashwantsahu20 replied:
fixed

renecannao commented on lib/MySQL_Session.cpp:
```diff
@@ -6414,6 +6415,54 @@ bool MySQL_Session::handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_C
+						}
+
+						wait_timeout = client_timeout;
+						proxy_debug(PROXY_DEBUG_MYSQL_COM, 8, "Changing connection wait_timeout to %llu ms\n", client_timeout);
```
Setting, more than changing.
The code doesn't actually check if there was a change

yashwantsahu20 replied:
fixed

renecannao commented on lib/MySQL_Session.cpp:
```diff
@@ -6414,6 +6415,54 @@ bool MySQL_Session::handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_C
+
+						// Warn if client's value exceeds current global timeout
+						if (client_timeout > mysql_thread___wait_timeout) {
+							proxy_error("[Warning] Client-specified wait_timeout (%llu ms) exceeds global mysql-wait_timeout (%llu ms). Global timeout will still be enforced.",
```
Did you mean to use `proxy_warning()` instead of `proxy_error()` ?

Also, it would be useful to log client's details (IP at least, perhaps username too).

yashwantsahu20 replied:
fixed

renecannao commented on lib/MySQL_Session.cpp:
```diff
@@ -6414,6 +6415,54 @@ bool MySQL_Session::handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_C
+						unsigned long long client_timeout = 0;
+						try {
+							client_timeout = std::stoull(value) * 1000;
+						} catch (const std::exception& e) {
```
I don't think input validation should be done here.
Numeric variables (like `wait_timeout`) should be defined in `mysql_variables_numeric` , and marked as numeric.
IIRC , the parser should take care of validating that the value is indeed numeric.

yashwantsahu20 replied:
Adding into mysql_variables_numeric doesnt work and exits.
As it seems mysql_variables_numeric is more related to tracking variable.

```
					} else if (mysql_variables_numeric.find(var) != mysql_variables_numeric.end()) {
						int idx = SQL_NAME_LAST_HIGH_WM;
						for (int i = 0 ; i < SQL_NAME_LAST_HIGH_WM ; i++) {
							if (mysql_tracked_variables[i].is_number) {
```

with the original fix it gives error message like direct MySQL server and It will par with MySQL server

yashwantsahu20 committed:
Removed range check.

renecannao commented:
Can one of the admins verify this patch?

wazir-ahmed committed:
Add TAP test for session wait timeout

Signed-off-by: Wazir Ahmed <wazir@proxysql.com>

wazir-ahmed committed:
fix: Close idle session based on effective timeout

- effective_timeout = min (mysql-wait_timeout, session-wait_timeout)

Signed-off-by: Wazir Ahmed <wazir@proxysql.com>

renecannao commented:
retest this please

renecannao committed:
Merge branch 'v3.0' into v3.0_wait_timeout

Signed-off-by: René Cannaò <rene@proxysql.com>

renecannao commented:
retest this please

renecannao committed:
session: Fix wait_timeout member variable declaration and usage

- Add wait_timeout member variable declaration to Base_Session class
- Fix constructor initialization to use this->wait_timeout
- Fix assignment in handler to properly scope member variable
- Resolves compilation error for wait_timeout functionality

renecannao committed:
session: Add input validation for client wait_timeout with silent clamping

- Add range validation for client SET wait_timeout commands
- Implement clamping between 1 second (1000ms) and 20 days (1,728,000,000ms)
- Add warning messages when values are clamped due to ProxySQL limits
- Maintain MySQL compatibility by accepting larger values than global config
- Fix signed/unsigned comparison warning in wait_timeout assignment
- Ensures client applications don't break while enforcing safety limits

renecannao committed:
session: Add wait_timeout to proxysql internal session JSON

- Include wait_timeout value in session JSON output for monitoring/debugging
- Provides visibility into client-configured timeout values

renecannao committed:
test: Add comprehensive wait_timeout JSON validation tests

- Add JSON parsing helper for PROXYSQL INTERNAL SESSION responses
- Add test_wait_timeout_json_values() function with 8 test cases:
  * Zero value (should clamp to 1 second)
  * Valid values (1s, 10s, 5min, 1hr, 24hr, 20 days)
  * Value exceeding 20 days (should clamp to maximum)
- Validate that wait_timeout appears in session JSON output
- Test plan increased from 4 to 12 tests total
- Provides comprehensive validation of wait_timeout clamping behavior

renecannao committed:
Fix mysql-set_wait_timeout-t.cpp compilation and JSON testing

- Replace <json/json.h> include with <json.hpp> and proper nlohmann namespace
- Replace manual JSON parsing with fetch_internal_session() utility function
- Replace fail() calls with ok(false, ...) to follow TAP framework standards
- Add comprehensive JSON validation test for various wait_timeout values
- Include edge case testing: zero values, maximum values, and clamping behavior

Resolves compilation issues and enables proper testing of wait_timeout JSON output.

renecannao committed:
Fix wait_timeout timeout calculations and add proper newline characters

Key improvements:
- Fix timeout comparison in MySQL_Thread::idle_thread_to_kill_idle_sessions() to prevent underflow
- Use effective wait_timeout (minimum of global and session values) for idle timeout calculations
- Add proper newline characters to proxy_warning messages for consistent log formatting
- Increase test sleep times to account for global timeout enforcement
- Fix session timeout test durations to properly test timeout behavior

Technical changes:
- Replace broken min_idle calculation with proper effective wait_timeout logic
- Add std::min() usage to determine effective timeout from global and session values
- Ensure warning messages end with newline characters for proper log formatting
- Update test sleep durations to ensure proper timeout testing

Resolves potential timeout calculation bugs and ensures consistent timeout enforcement behavior.

renecannao committed:
Refactor idle session scanning and improve test precision

Code improvements:
- Extract SESS_TO_SCAN_idle_thread constant to header file for better maintainability
- Replace magic number 128 with named constant in idle_thread_to_kill_idle_sessions()
- Improve code readability and consistency in session scanning logic

Test enhancements:
- Add mysql-poll_timeout configuration for more precise timeout testing
- Reduce test sleep times to 13 seconds for faster test execution
- Add diagnostic messages to clearly show timeout configurations in test output
- Ensure tests properly validate timeout enforcement with precise timing

The changes improve code maintainability and make tests more reliable and faster
while maintaining accurate timeout validation.

renecannao committed:
Improve wait_timeout warning messages with detailed connection information

Enhance logging clarity:
- Replace generic IP address with detailed connection info including IP and port
- Use client_myds->addr.addr and client_myds->addr.port for precise identification
- Improve debuggability of timeout clamping and enforcement warnings

The warning messages now provide complete connection details, making it easier
to identify and troubleshoot timeout-related issues in ProxySQL logs.

renecannao committed:
Fix metrics collection for wait_timeout counters

The get_status_variable() function was only scanning worker threads
but ignoring auxiliary threads (idle threads) where timeout
terminations are detected. This caused the timeout termination
counter to show incorrect/zero values.

- Added idle thread scanning to both overloaded versions of
  get_status_variable() function
- Now properly collects metrics from both worker and idle threads
- Fixes the issue where proxysql_mysql_timeout_terminated_connections_total
  showed zero despite actual timeout terminations

Resolves the metrics reading issue identified in the previous commits.

renecannao committed:
Update test groups configuration
